## Data analysis of Airbnb NYC 2019 data set
#### Umesh Patel // 20220329

### To understand Airbnb New York 2019 dataset

### Introduction

### Importing libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
# %matplotlib inline
# plt.style.use(‘seaborn-dark-palette’)
# from scipy import stats
# import datetime as dt
# import plotly
# import plotly.express as px

### Ignore warnings

In [2]:
import warnings
warnings.filterwarnings('ignore')
# pd.set_option(‘display.max_columns’, None)
# pd.set_option(‘display.max_rows’, None)

### Loading data

In [3]:
airbnb = pd.read_csv("AirbnbNYC2019.csv")

### Reviewing data

In [4]:
airbnb.head(3)

id                                 name  host_id  host_name  \
0  2539   Clean & quiet apt home by the park     2787       John   
1  2595                Skylit Midtown Castle     2845   Jennifer   
2  3647  THE VILLAGE OF HARLEM....NEW YORK !     4632  Elisabeth   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   
2    150               3                  0         NaN                NaN   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365

In [5]:
airbnb.tail(3)

id                                               name   host_id  \
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group   neighbourhood  latitude  longitude  \
48892  Ilgar & Aysel           Manhattan          Harlem  40.81475  -73.94867   
48893            Taz           Manhattan  Hell's Kitchen  40.75751  -73.99112   
48894     Christophe           Manhattan  Hell's Kitchen  40.76404  -73.98933   

             room_type  price  minimum_nights  number_of_reviews last_review  \
48892  Entire home/apt    115              10                  0         NaN   
48893      Shared room     55               1                  0         NaN   
48894     Private room     90               7                  0         NaN   

       reviews_per_month  calculated_host_listings_count  availability_365  
48892                NaN                               1                27  
48893                NaN                               6                 2  
48894                NaN                               1                23

In [6]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [7]:
airbnb.shape

(48895, 16)

In [8]:
airbnb.index

RangeIndex(start=0, stop=48895, step=1)

In [10]:
airbnb.nunique()

id                                48895
name                              47905
host_id                           37457
host_name                         11452
neighbourhood_group                   5
neighbourhood                       221
latitude                          19048
longitude                         14718
room_type                             3
price                               674
minimum_nights                      109
number_of_reviews                   394
last_review                        1764
reviews_per_month                   937
calculated_host_listings_count       47
availability_365                    366
dtype: int64

#### Converting dtypes: Most of the columns do not have proper data types so I will convert the data types of these columns

* id, host_id: These columns are int64 but won't be used for any math calculations so I will convert them to string.

* neighbourhood_group, neighbourhood: These columns are object but I will convert them to string because they are names.

* latitude, longitude: float16 can handle these numerical data so I will change them from float64 to float16.

* room_type: This is object and has only 3 distinct values so I will convert to category

* price: Price can be a few thousands $ at the most so I will convert it from int64 to int16.

* minimum_nights: int16 can handle this data so I will convert it from int64 to int16.

* number_of_reviews: This doesn't need an int64 so will convert it to int16 which is enough for this parameter.

* last_review: This is object instead of date so will convert it to datetime64.

* reviews_per_month: float16 can accomodate data in this column so I will convert it from float64 to float16.

* calculated_host_listings_count: int16 can accomodate listings/host so I will convert it from int64 to int16.

* availability_365: I will convert this from int64 to int16 because it can't be larger than 366 and int16 would be OK.

In [11]:
# Cconverting dtypes

airbnb = airbnb.astype({'id':'string',
                        'name':'string',
                        'host_id':'string',
                        'host_name':'string',
                        'neighbourhood_group':'string',
                        'neighbourhood':'string',
                        'latitude':'float16',
                        'longitude':'float16',
                        'room_type':'category',
                        'price':'int16',
                        'minimum_nights':'int16',
                        'number_of_reviews':'int16',
                        'last_review':'datetime64',
                        'reviews_per_month':'float16',
                        'calculated_host_listings_count':'int16',
                        'availability_365':'int16'
                        })

In [12]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48895 non-null  string        
 1   name                            48879 non-null  string        
 2   host_id                         48895 non-null  string        
 3   host_name                       48874 non-null  string        
 4   neighbourhood_group             48895 non-null  string        
 5   neighbourhood                   48895 non-null  string        
 6   latitude                        48895 non-null  float16       
 7   longitude                       48895 non-null  float16       
 8   room_type                       48895 non-null  category      
 9   price                           48895 non-null  int16         
 10  minimum_nights                  48895 non-null  int16         
 11  nu

### Exploring data (manipulation)

#### Renaiming columns

'calculated_host_listings_count' column name is too long and it is confusing with the word 'calculated.' It is about listings per host so I will change it to 'host_listings_count.'

In [13]:
airbnb = airbnb.rename(columns = {'calculated_host_listings_count':'host_listings_count'})

In [14]:
airbnb.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'host_listings_count', 'availability_365'],
      dtype='object')

#### Removing unnecessary columns

* "name" column in this data has short description / comment about the property which is of not much use for the analysis so I will remove this column.
* "host_name" column contains names of the property owners and "host_id" contains IDs of the owners. I need only one column so will remove "host_name" column.

In [45]:
airbnb = airbnb.drop(columns = ['name', 'host_name'])

In [46]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id                   48895 non-null  string        
 1   host_id              48895 non-null  string        
 2   neighbourhood_group  48895 non-null  string        
 3   neighbourhood        48895 non-null  string        
 4   latitude             48895 non-null  float16       
 5   longitude            48895 non-null  float16       
 6   room_type            48895 non-null  category      
 7   price                48895 non-null  int16         
 8   minimum_nights       48895 non-null  int16         
 9   number_of_reviews    48895 non-null  int16         
 10  last_review          38843 non-null  datetime64[ns]
 11  reviews_per_month    38843 non-null  float16       
 12  host_listings_count  48895 non-null  int16         
 13  availability_365     48895 non-

By changing the data types, I have reduced the size of airbnb dataframe from 6.0 MB to 2.7 MB which is more than 50% reduction.

Checking duplicate rows

In [48]:
airbnb.duplicated().value_counts()

False    48895
dtype: int64

There are no duplicate rows.

In [49]:
airbnb.describe()

latitude   longitude         price  minimum_nights  \
count  4.889500e+04  48895.0000  48895.000000    48895.000000   
mean            inf        -inf    152.720687        7.029962   
std    5.590820e-02      0.0513    240.154170       20.510550   
min    4.050000e+01    -74.2500      0.000000        1.000000   
25%    4.068750e+01    -74.0000     69.000000        1.000000   
50%    4.071875e+01    -73.9375    106.000000        3.000000   
75%    4.075000e+01    -73.9375    175.000000        5.000000   
max    4.090625e+01    -73.6875  10000.000000     1250.000000   

       number_of_reviews  reviews_per_month  host_listings_count  \
count       48895.000000       38843.000000         48895.000000   
mean           23.274466           1.372070             7.143982   
std            44.550582           1.680664            32.952519   
min             0.000000           0.010002             1.000000   
25%             1.000000           0.189941             1.000000   
50%             5.000000           0.720215             1.000000   
75%            24.000000           2.019531             2.000000   
max           629.000000          58.500000           327.000000   

       availability_365  
count      48895.000000  
mean         112.781327  
std          131.622289  
min            0.000000  
25%            0.000000  
50%           45.000000  
75%          227.000000  
max          365.000000

Finding missing data in each column

In [67]:
airbnb.isna().sum()

id                         0
host_id                    0
neighbourhood_group        0
neighbourhood              0
latitude                   0
longitude                  0
room_type                  0
price                      0
minimum_nights             0
number_of_reviews          0
last_review            10052
reviews_per_month      10052
host_listings_count        0
availability_365           0
dtype: int64

In [68]:
airbnb['last_review'].value_counts(dropna = False, normalize = True)

NaT           0.205583
2019-06-23    0.028899
2019-07-01    0.027794
2019-06-30    0.027426
2019-06-24    0.017895
                ...   
2012-12-25    0.000020
2013-10-01    0.000020
2014-05-29    0.000020
2014-04-19    0.000020
2018-03-29    0.000020
Name: last_review, Length: 1765, dtype: float64

In [57]:
airbnb['reviews_per_month'].value_counts(dropna = False)

NaN          10052
0.020004       919
1.000000       893
0.049988       893
0.029999       804
             ...  
9.531250         1
9.742188         1
6.058594         1
8.250000         1
10.539062        1
Name: reviews_per_month, Length: 938, dtype: int64

Finding missing data in each row -- how many columns (in each row) have True or missing values

In [59]:
airbnb.isna().sum(axis = 1)

0        0
1        0
2        2
3        0
4        0
        ..
48890    2
48891    2
48892    2
48893    2
48894    2
Length: 48895, dtype: int64

In [60]:
num_missing_by_row = airbnb.isna().sum(axis = 1)

In [65]:
(num_missing_by_row > 0).sum()

10052

In [63]:
airbnb[num_missing_by_row > 0]

id   host_id neighbourhood_group       neighbourhood  latitude  \
2          3647      4632           Manhattan              Harlem  40.81250   
19         7750     17985           Manhattan         East Harlem  40.78125   
26         8700     26394           Manhattan              Inwood  40.87500   
36        11452      7355            Brooklyn  Bedford-Stuyvesant  40.68750   
38        11943     45445            Brooklyn            Flatbush  40.62500   
...         ...       ...                 ...                 ...       ...   
48890  36484665   8232441            Brooklyn  Bedford-Stuyvesant  40.68750   
48891  36485057   6570630            Brooklyn            Bushwick  40.68750   
48892  36485431  23492952           Manhattan              Harlem  40.81250   
48893  36485609  30985759           Manhattan      Hell's Kitchen  40.75000   
48894  36487245  68119814           Manhattan      Hell's Kitchen  40.75000   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
2       -73.9375     Private room    150               3                  0   
19      -73.9375  Entire home/apt    190               7                  0   
26      -73.9375     Private room     80               4                  0   
36      -73.9375     Private room     35              60                  0   
38      -73.9375     Private room    150               1                  0   
...          ...              ...    ...             ...                ...   
48890   -73.9375     Private room     70               2                  0   
48891   -73.9375     Private room     40               4                  0   
48892   -73.9375  Entire home/apt    115              10                  0   
48893   -74.0000      Shared room     55               1                  0   
48894   -74.0000     Private room     90               7                  0   

      last_review  reviews_per_month  host_listings_count  availability_365  
2             NaT                NaN                    1               365  
19            NaT                NaN                    2               249  
26            NaT                NaN                    1                 0  
36            NaT                NaN                    1               365  
38            NaT                NaN                    1               365  
...           ...                ...                  ...               ...  
48890         NaT                NaN                    2                 9  
48891         NaT                NaN                    2                36  
48892         NaT                NaN                    1                27  
48893         NaT                NaN                    6                 2  
48894         NaT                NaN                    1                23  

[10052 rows x 14 columns]

For time series analysis there is only last_review column whch won't be much useful so I am adding year, month, day
and day_name columns after extracting them from the last_review column. I am also changing the data types of these
new columns.

In [14]:
airbnb['year'] = airbnb['last_review'].dt.year.astype('float16')
airbnb['month'] = airbnb['last_review'].dt.month.astype('float16')
airbnb['day'] = airbnb['last_review'].dt.day.astype('float16')
airbnb['day_name'] = airbnb['last_review'].dt.day_name().astype('string')

In [15]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48895 non-null  string        
 1   host_id                         48895 non-null  string        
 2   neighbourhood_group             48895 non-null  string        
 3   neighbourhood                   48895 non-null  string        
 4   latitude                        48895 non-null  float16       
 5   longitude                       48895 non-null  float16       
 6   room_type                       48895 non-null  category      
 7   price                           48895 non-null  int16         
 8   minimum_nights                  48895 non-null  int16         
 9   number_of_reviews               48895 non-null  int16         
 10  last_review                     38843 non-null  datetime64[ns]
 11  re

#### Exploring data (manipulation)

In [16]:
airbnb.head(3)

id host_id neighbourhood_group neighbourhood  latitude  longitude  \
0  2539    2787            Brooklyn    Kensington  40.65625   -74.0000   
1  2595    2845           Manhattan       Midtown  40.75000   -74.0000   
2  3647    4632           Manhattan        Harlem  40.81250   -73.9375   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaT   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0           0.209961                               6               365   
1           0.379883                               2               355   
2                NaN                               1               365   

     year  month   day day_name  
0  2018.0   10.0  19.0   Friday  
1  2019.0    5.0  21.0  Tuesday  
2     NaN    NaN   NaN     <NA>

In [17]:
# selecting all numeric columns
airbnb_num = airbnb.select_dtypes(include = 'number').columns
airbnb_num

Index(['latitude', 'longitude', 'price', 'minimum_nights', 'number_of_reviews',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365', 'year', 'month', 'day'],
      dtype='object')

In [18]:
# selecting all non-numeric columns
airbnb_no_num = airbnb.select_dtypes(exclude = 'number').columns
airbnb_no_num

Index(['id', 'host_id', 'neighbourhood_group', 'neighbourhood', 'room_type',
       'last_review', 'day_name'],
      dtype='object')

In [19]:
# Analyzing numerical values
airbnb.select_dtypes(include = 'number').describe()

latitude   longitude         price  minimum_nights  \
count  4.889500e+04  48895.0000  48895.000000    48895.000000   
mean            inf        -inf    152.720687        7.029962   
std    5.590820e-02      0.0513    240.154170       20.510550   
min    4.050000e+01    -74.2500      0.000000        1.000000   
25%    4.068750e+01    -74.0000     69.000000        1.000000   
50%    4.071875e+01    -73.9375    106.000000        3.000000   
75%    4.075000e+01    -73.9375    175.000000        5.000000   
max    4.090625e+01    -73.6875  10000.000000     1250.000000   

       number_of_reviews  reviews_per_month  calculated_host_listings_count  \
count       48895.000000       38843.000000                    48895.000000   
mean           23.274466           1.372070                        7.143982   
std            44.550582           1.680664                       32.952519   
min             0.000000           0.010002                        1.000000   
25%             1.000000           0.189941                        1.000000   
50%             5.000000           0.720215                        1.000000   
75%            24.000000           2.019531                        2.000000   
max           629.000000          58.500000                      327.000000   

       availability_365         year        month           day  
count      48895.000000  38843.00000  38843.00000  3.884300e+04  
mean         112.781327          inf          inf           inf  
std          131.622289      1.21582      2.53125  9.882812e+00  
min            0.000000   2011.00000      1.00000  1.000000e+00  
25%            0.000000   2018.00000      5.00000  6.000000e+00  
50%           45.000000   2019.00000      6.00000  1.700000e+01  
75%          227.000000   2019.00000      7.00000  2.400000e+01  
max          365.000000   2019.00000     12.00000  3.100000e+01

In [20]:
# Analyzing categorical values
airbnb.select_dtypes(exclude = 'number').describe()

id    host_id neighbourhood_group neighbourhood        room_type  \
count   48895      48895               48895         48895            48895   
unique  48895      37457                   5           221                3   
top      2539  219517861           Manhattan  Williamsburg  Entire home/apt   
freq        1        327               21661          3920            25409   
first     NaN        NaN                 NaN           NaN              NaN   
last      NaN        NaN                 NaN           NaN              NaN   

                last_review day_name  
count                 38843    38843  
unique                 1764        7  
top     2019-06-23 00:00:00   Sunday  
freq                   1413     9382  
first   2011-03-28 00:00:00      NaN  
last    2019-07-08 00:00:00      NaN